lib/ansible/modules/identity/keycloak/keycloak_group.py

```diff
@@ -0,0 +1,358 @@
+
+    # prepare the new group
+    updated_group = before_group.copy()
+    updated_group.update(changeset)
+
+    result['proposed'] = changeset if state != 'absent' else {}
+    result['existing'] = before_group
+
+    # if before_group is none, the group doesn't exist.
+    if before_group == {}:
+        if state == 'absent':
+            # nothing to do.
+            if module._diff:
+                result['diff'] = dict(before='', after='')
+            result['msg'] = 'Group does not exist; doing nothing.'
+            result['end_state'] = dict()
+            module.exit_json(**result)
+
+        # for 'present', create a new group.
+        result['changed'] = True
+        if name is None:
+            module.fail_json(msg='name must be specified when creating a new group')
+
+        if module._diff:
+            result['diff'] = dict(before='', after=after_group)
+
+        if module.check_mode:
+            module.exit_json(**result)
+
+        # do it for real!
+        kc.create_group(updated_group, realm=realm)
+        after_group = kc.get_group_by_name(name, realm)
+
+        result['end_state'] = after_group
+        result['msg'] = 'Group {name} has been created with ID {id}'.format(name=after_group['name'],
+                                                                            id=after_group['id'])
+
+    else:
+        if state == 'present':
+            # no changes
+            if updated_group == before_group:
+                result['changed'] = False
+                result['end_state'] = updated_group
+                result['msg'] = "No changes required to group {name}.".format(name=before_group['name'])
+                module.exit_json(**result)
+
+            # update the existing group
+            result['changed'] = True
+
+            if module._diff:
+                result['diff'] = dict(before=before_group, after=updated_group)
+
+            if module.check_mode:
+                module.exit_json(**result)
+
+            # do the update
+            kc.update_group(updated_group, realm=realm)
+
+            after_group = kc.get_group_by_groupid(updated_group['id'], realm=realm)
+
+            result['end_state'] = after_group
+            result['msg'] = "Group {id} has been updated".format(id=after_group['id'])
+
+            module.exit_json(**result)
+
+        elif state == 'absent':
+            result['proposed'] = dict()
+            result['end_state'] = dict()
+
+            if module._diff:
+                result['diff'] = dict(before=before_group, after='')
+
+            if module.check_mode:
+                module.exit_json(**result)
+
+            # delete for real
+            gid = before_group['id']
+            kc.delete_group(groupid=gid, realm=realm)
+
+            result['changed'] = True
+            result['msg'] = "Group {name} has been deleted".format(name=before_group['name'])
+
+            module.exit_json(**result)
+
+        else:
```

Comment:
This branch can never be hit since the argument_spec defines which states are valid and checks the provided module arguments against it.

Reply:
Good point - just me being overcautious. Will remove.